sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,373 @@
+            posterior is weighted with 1/N.
+        default_x: Used in `.sample(), .log_prob()` as default conditioning context.
+    """
+
+    def __init__(self, posteriors: List, weights: Optional[List] or Tensor = None):
+        """
+
+        Args:
+            posteriors: List containing the trained posterior instances that will make
+                up the ensemble.
+            weights: Assign weights to posteriors manually, otherwise they will be
+                weighted with 1/N.
+        """
+        self.posteriors = posteriors
+        self.num_components = len(posteriors)
+        self.weights = weights
+        self.potential_fn = None
+
+    @property
+    def weights(self):
+        return self._weights
+
+    @weights.setter
+    def weights(self, weights: List[float] or Tensor):
+        """Set relative weight for each posterior in the ensemble.
+
+        Weights are normalised.
+
+        Args:
+            weights: Assignes weight to each posterior distribution.
+        """
+        if weights is None:
+            self._weights = torch.tensor(
+                [1.0 / self.num_components for _ in range(self.num_components)]
+            )
+        elif type(weights) == Tensor or type(weights) == List:
+            self._weights = torch.tensor(weights) / sum(weights)
+        else:
+            raise TypeError
+
+    def sample(self, sample_shape: Shape = torch.Size(), **kwargs) -> Tensor:
+        r"""Return samples from posterior ensemble.
+
+        All kwargs are passed directly through to `posterior.sample()`.
+
+        Args:
+            sample_shape: Desired shape of samples that are drawn from posterior
+                ensemble. If sample_shape is multidimensional we simply draw
+                    `sample_shape.
+            numel()` samples and then reshape into the desired shape.
+        """
+        num_samples = torch.Size(sample_shape).numel()
+        idxs = torch.multinomial(self._weights, num_samples, replacement=True)
+        samples = []
+        for c, n in torch.vstack(idxs.unique(return_counts=True)).T:
+            sample_shape_c = torch.Size((n,))
+            samples.append(self.posteriors[c].sample(sample_shape_c, **kwargs))
+        return torch.vstack(samples).reshape(*sample_shape, -1)
+
+    def log_prob(self, theta: Tensor, individually: bool = False, **kwargs) -> Tensor:
+        r"""Returns the average log-probability of the posterior ensemble
+
+        $\sum_{i}^{N} w_{i} p_i(\theta|x)$.
+
+        All kwargs are passed directly through to `posterior.log_prob()`.
+
+        Args:
+            theta: Parameters $\theta$.
+            individually: If true, returns log weights and log_probs individually.
+
+        Returns:
+            `(len(θ),)`-shaped average log posterior probability $\log p(\theta|x)$ for
+            θ in the support of the prior, -∞ (corresponding to 0 probability) outside.
+        """
```

Comment:
`log_prob` does not make much sense when different methods are used (because SNPE posteriors are normalized but SNLE posteriors are not). Maybe raise an error if self.posteriors are of different types?